src/lib.rs

```diff
@@ -139,123 +135,190 @@ pub struct Options {
 
-fn snip_nth(
-    n: usize,
-    num_imports: usize,
-    code: &mut elements::CodeSection,
-) -> Result<(), failure::Error> {
-    if n < num_imports {
-        bail!("cannot snip imported functions");
+                let unreachable = builder.unreachable();
+                *expr_id = builder.with_side_effects(dropped_args, unreachable, vec![]);
+            }
+
+            (*expr_id).visit_mut(self);
+        }
     }
 
-    let body = code.bodies_mut()
-        .get_mut(n - num_imports)
-        .ok_or(failure::err_msg(format!(
-            "index {} is out of bounds of the code section",
-            n - num_imports
-        )))?;
+    module.funcs.par_iter_local_mut().for_each(|(id, func)| {
+        // Don't bother transforming functions that we are snipping.
+        if to_snip.contains(&id) {
+            return;
+        }
+
+        let mut entry = func.entry_block();
+        let v = &mut Replacer { func, to_snip };
+        v.visit_block_id_mut(&mut entry);
+    });
+}
 
-    *body.code_mut().elements_mut() = vec![elements::Instruction::Unreachable, elements::Instruction::End];
+fn unexport_snipped_functions(module: &mut walrus::Module, to_snip: &HashSet<walrus::FunctionId>) {
+    let exports_to_snip: HashSet<walrus::ExportId> = module
+        .exports
+        .iter()
+        .filter_map(|e| match e.item {
+            walrus::ExportItem::Function(f) if to_snip.contains(&f) => Some(e.id()),
+            _ => None,
+        })
+        .collect();
 
-    Ok(())
+    for e in exports_to_snip {
+        module.exports.delete(e);
+    }
 }
 
-trait NamesSectionNames {
-    fn names_section_names(&self) -> Option<elements::NameMap>;
+fn unimport_snipped_functions(module: &mut walrus::Module, to_snip: &HashSet<walrus::FunctionId>) {
+    let imports_to_snip: HashSet<walrus::ImportId> = module
+        .imports
+        .iter()
+        .filter_map(|i| match i.kind {
+            walrus::ImportKind::Function(f) if to_snip.contains(&f) => Some(i.id()),
+            _ => None,
+        })
+        .collect();
+
+    for i in imports_to_snip {
+        module.imports.delete(i);
+    }
 }
 
-impl NamesSectionNames for elements::Module {
-    fn names_section_names(&self) -> Option<elements::NameMap> {
-        for section in self.sections() {
-            if let &elements::Section::Name(elements::NameSection::Function(ref name_section)) =
-                section
-            {
-                return Some(name_section.names().clone());
-            }
+fn snip_table_elements(module: &mut walrus::Module, to_snip: &HashSet<walrus::FunctionId>) {
+    let mut builder = walrus::FunctionBuilder::new();
+    let unreachable = builder.unreachable();
+    let ty = module.types.add(&[], &[]);
+    let unreachable_func = builder.finish(ty, vec![], vec![unreachable], module);
+
+    for t in module.tables.iter_mut() {
+        if let walrus::TableKind::Function(ref mut ft) = t.kind {
+            ft.elements
+                .par_iter_mut()
```

Comment:
The parallel here and below may be a bit overkill

Reply:
rayon should just not split it out to too many workers if it doesn't make sense, right? Or at least if it isn't going to take enough time to notice either way, then using rayon doesn't hurt...

Reply:
Fixed this too, since the previous fix made it so that this had to mutably borrow parts of the module, which means it wasn't safe for parallelism anymore anyways.